### ISSUE TITLE
Incorrect syntax on queries

### ISSUE THREAD

**Issue by erikest** [ProjectNami/projectnami#98]
I'm getting a few errors that seem to be related to query translation:

one takes the form:

SELECT user_id FROM wp_wysija_user WHERE email LIKE"support@foo.com" ORDER BY user_id DESC OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY

with the error:
"Invalid Column Name 'support@foo.com'

here the issue is the use of double quotes for the like operator

another has the form:
SELECT \* FROM wp_wysija_user_list WHERE user_id=1 AND list_id=2 OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY

and error:
Incorrect syntax near 'OFFSET'

here the issue is the omission of ORDER BY before the OFFSET.  It looks like the queries are getting passed through directly without translation.

another is :
[[Microsoft][ODBC Driver 11 for SQL Server][SQL Server]Incorrect syntax near '|'. Code - 102]
SELECT t.tid, t.stub, p.stub as parent_stub, p.topic_parent as parent_parent, p.topic_subject as parent_subject, p.topic_category as parent_category, t.topic_subject, t.topic_owner, t.topic_post, t.topic_started, t.topic_category, t.topic_date, u.display_name, t.topic_parent, t.topic_group FROM wp_symposium_topics t INNER JOIN wp_users u ON t.topic_owner = u.ID LEFT JOIN wp_symposium_topics p ON t.topic_parent = p.tid WHERE t.topic_approved = 'on' AND (t.topic_parent = 0 || p.topic_parent = 0) AND t.topic_owner != 0 ORDER BY t.tid DESC OFFSET 0 ROWS FETCH NEXT 100 ROWS ONLY

which looks like it's a mysql query, since it is using || as a logical OR.

These all happen when I update a user profile in Symposium.

I've searched through the MailPoet source a bit, but can't figure out where those queries are coming from.  I thought ProjectNami was supposed to convert these queries at a level below the plugin, so the plugin doesn't have any awareness of the db platform.  Is this an edge case?

Thanks for any help!

**Comment by patrickebates** [comment]
You are correct, Project Nami will convert queries if the queries are using the standard WP APIs.  In this case, though, it seems the plugin is making a direct DB call with MySQL syntax.  If we can locate the source of the call, we should be able to prepare a custom translation.

**Comment by erikest** [comment]
Well, hmm that last one is the most concerning then, because I'm just realizing my client has stopped using the MailPoet plugin prior to this move, so I'm just going to uninstall it - don. The last one though appears on a side bar and is clearly from Symposium.  So what you're suggesting is I hunt down that code in the source somewhere and replace the || with an OR?  Obviously that still leaves me vulnerable to updates rebreakifying it...  Hmm, should I start complaining to Symposium that they're not using the 'standard WP APIs'? :)  

Searching through the Symposium code base is not making me feel good - they have tons of hard coded sql, but they are using a global $wbdp, like $wpdb->get_results($wpdb->prepare($sql, $id)).  Is there a hook I can get on the prepare or get_results that will allow me to do a simple replace of || with OR?  Am I on the right track || grabbing at straws?

Thanks for quick reply

**Comment by patrickebates** [comment]
If you want to do the replace as a stopgap, that works.  But the long term solution will be an addition to /wp-includes/translations.php to handle this specific case.

**Comment by erikest** [comment]
Ok, that's the extension point.  So I should do the change and make a pull
request, or were you implying that you guys would look at it?

**Comment by patrickebates** [comment]
If you'd like to make a pull request, that will probably be faster than waiting for us to look at it.  Besides, we're not familiar with the plugin in question either.

**Comment by erikest** [comment]
What? You guys don't know every WordPress plugin?  :) I'm shocked.

Well, in this case, it should be easy enough since the issue seems like a
simple replace..  I'm on it!  After this cup of coffee..

Thank you for the pointers :)

**Comment by erikest** [comment]
Maybe I can get your thoughts on a couple of edge cases I'm considering as I add a translate function on my branch.

First, it looks like MySQL allows &&, || and ! for AND, OR and NOT, so I'd like to try to handle those three cases.  the ! will be a little trickier because TSQL allows !=, !< and !>.  So I'll test that the ! isn't followed by any of those three characters, but the other issue with this case is that ! could exist in string literals for a where/like condition.  I'm no regexpert, so crafting an expression that will correctly determine if the bang is inside a string might be something you just know off hand...  This issue technically applies to the && and || operators, but maybe I'm too worried about these edge cases?

The problem I see with just testing for && between two quotes is a condition of the form:
WHERE colA = 'foo' && colB = 'bar'.  I'm thinking I need something that will express if there are an odd number of single quotes preceding the &&, to indicate an non-terminated string.

If you've any thoughts as I work my way through this, 'that'd be greaaat' :)

**Comment by patrickebates** [comment]
My personal rule is not to be too broad when working on a plugin translation, but instead be as specific as possible.  So maybe search for `(t.topic_parent = 0 || p.topic_parent = 0)` and replace with `(t.topic_parent = 0 OR p.topic_parent = 0)`

**Comment by patrickebates** [comment]
I believe this case has become a microcosm of why Project Nami exists as a full port rather than simply relying on a wall of regex to convert the syntax.

**Comment by erikest** [comment]
Ya, catching all the cases is tough business :)  

Too bad the query couldn't be passed through SSMA for MySQL, which folks say does a pretty good job of converting statements.  It did a *pretty* good job of bringing my database over on the initial import.  This doc [here](https://msdn.microsoft.com/en-us/library/hh313124(v=sql.110).aspx) seems to indicate that you can convert statements via the console and an input file.  If the statement could be passed through that on its way to your mssql layer, it would probably catch most things.  Who knows how performant that approach would be though.

Seems like plugins using hardcoded queries is par for the course, so that wall of regex or something is always going to need to be there...

**Comment by patrickebates** [comment]
Is that one replacement all that is needed to make the plugin operate?

**Comment by erikest** [comment]
Well, it's the only problem I've run into so far and it only occurred when
saving a forum users profile.  I can search through the source though and
figure out if they use the same pattern elsewhere.  I thought, if I ran
into other issues, I could just expand that function further...

The other query errors I mentioned initially were due to mail poet, which
we didn't actually need, but I could include replaces for those as well,
since there are some 300k users of that plugin, so someone is likely to run
into those issues in the future.
